File: sol/solution/0001-0100/0022/Solution.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> result;
        function<void(string, int, int)> backtrack = [&](string curr, int open, int close) {
            if (curr.size() == n * 2) {
                result.push_back(curr);
                return;
            }
            if (open < n)
                backtrack(curr + '(', open + 1, close);
            if (close < open)
                backtrack(curr + ')', open, close + 1);
        };
        backtrack("", 0, 0);
        return result;
    }
};

File: sol/solution/0001-0100/0022/Solution.scala
object Solution {
    def generateParenthesis(n: Int): List[String] = {
        def backtrack(s: String, left: Int, right: Int): List[String] = {
            if (s.length == n * 2) List(s)
            else {
                val addLeft = if (left < n) backtrack(s + "(", left + 1, right) else Nil
                val addRight = if (right < left) backtrack(s + ")", left, right + 1) else Nil
                addLeft ++ addRight
            }
        }
        backtrack("", 0, 0)
    }
}